File: mqtt/src/main/java/org/sheedon/mqtt/NamedRunnable.java
package org.sheedon.mqtt;

/**
 * Runnable implementation which always sets its thread name.
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2020/2/11 20:59
 */
public abstract class NamedRunnable implements Runnable {
    private final String name;

    NamedRunnable(String format, Object... args) {
        this.name = Util.format(format, args);
    }

    @Override public final void run() {
        String oldName = Thread.currentThread().getName();
        Thread.currentThread().setName(name);
        try {
            execute();
        } finally {
            Thread.currentThread().setName(oldName);
        }
    }

    protected abstract void execute();
}

File: mqtt/src/main/kotlin/org/sheedon/mqtt/internal/concurrent/NamedRunnable.kt
package org.sheedon.mqtt.internal.concurrent

import java.util.*

/**
 * 设置其线程名称的 Runnable 实现。
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2022/1/8 4:09 下午
 */
abstract class NamedRunnable protected constructor(format: String, vararg args: Any?) : Runnable {
    private val name: String = String.format(Locale.US, format, *args)

    override fun run() {
        val oldName = Thread.currentThread().name
        Thread.currentThread().name = name
        try {
            execute()
        } finally {
            Thread.currentThread().name = oldName
        }
    }

    protected abstract fun execute()

}